client/src/index.js

```diff
@@ -281,116 +244,99 @@ function createFS(options) {
+        if("onunload" in global){
+          global.addEventListener('unload', cleanupManager);
+        }
+
+        // Deals with race conditions if a sync was
+        // started immediately after connecting before
+        // this callback could be triggered
+        if(sync.state !== sync.SYNC_SYNCING) {
+          sync.state = sync.SYNC_CONNECTED;
+          sync.emit('connected');
+          log.info('MakeDrive connected to server');
+        }
 
-        // Wait on initial downstream sync events to complete
-        sync.onSyncing = function() {
-          // do nothing, wait for onCompleted()
-          log.info('Starting initial downstream sync');
-          startTime = Date.now();
+        sync.onSyncing = function(path) {
+          // A downstream sync might have just started,
+          // update the queue
+          downstreamQueue = manager.downstreams;
+          sync.state = sync.SYNC_SYNCING;
+          sync.emit('syncing', 'Sync started for ' + path);
+          log.info('Sync started for ' + path);
         };
-        sync.onCompleted = function(paths, needUpstream) {
-          // Downstream sync is done, finish connect() setup
-          var duration = (Date.now() - startTime) + 'ms';
-          log.info('Completed initial downstream sync in ' + duration);
-          downstreamSyncCompleted(paths, needUpstream);
+
+        sync.onCompleted = function(path, needsUpstream) {
+          // A downstream sync might have completed,
+          // update the queue
+          downstreamQueue = manager.downstreams;
+          needsUpstream = needsUpstream || [];
+
+          fs.appendPathsToSync(needsUpstream, function(err) {
+            if(err) {
+              sync.emit('error', err);
+              log.error('Error appending paths to upstream after sync completed for ' + path + ' with error', err);
+              return;
+            }
+
+            manager.needsUpstream = manager.needsUpstream.filter(function(upstreamPath) {
+              return needsUpstream.indexOf(upstreamPath) === -1;
+            });
+
+            fs.getPathsToSync(function(err, pathsToSync) {
+              var syncsLeft;
+
+              if(err) {
+                sync.emit('error', err);
+                log.error('Error retrieving paths to sync after sync completed for ' + path + ' with error', err);
+                return;
+              }
+
+              syncsLeft = pathsToSync ? pathsToSync.concat(downstreamQueue) : downstreamQueue;
+
+              sync.emit('completed', path);
+              log.info('Sync completed for ' + path);
+
+              if(!syncsLeft.length) {
+                sync.allCompleted();
+              }
+            });
+          });
+        };
+
+        // This is called when all nodes have been synced
+        // upstream and all downstream syncs have completed
+        sync.allCompleted = function() {
+          if(sync.state !== sync.SYNC_DISCONNECTED) {
+            // Reset the state if it was not done already
```

Comment:
sorry, the "if it was not done already" part is incorrect. It is not done anywhere else :P